woocommerce/google-listings-and-ads
# Adjust event tracking for the creating and claim buttons of Google Ads account

### Changes proposed in this Pull Request:

PT: pcTzPl-2iK-p2

Since the `context` property of most event tracking is currently hardcoded, and there is no proper way to obtain state or data from far away components when sending event tracking, this PR tries to implement it via `@wordpress/hooks`. The goal is to avoid passing event properties in a prop drilling fashion, e.g. `context` and `step` at the `<Stepper>` layer in this PR need to be sent together in buttons several layers away.

Due to the number of files being changed, to avoid this PR getting too large, I would like to use two tracking events to show how the mechanism works, and then add the remaining event tracking in a subsequent PR.

#### 📌 Mechanism:

- Add a React hook `useEventPropertiesFilter` to handle event properties across multiple layers of components via `@wordpress/hooks`.

#### 📌 Event tracking:

- Add `context` and `step` event properties to Extension and Ads onboarding steppers for use by their subcomponents.
- Add event tracking to `ClaimAccountButton`.
- Add `context` and `step` event properties to the click event tracking in `TermsModal`.

#### 📌 Extra change:

- Reuse `ClaimAccountButton` to reduce the duplicate implementation.
- Adjust jest config and setup so that tests related to event tracking don't have to validate global event properties.

### Screenshots:

#### 📷 Event tracking when clicking on button to create Google Ads account in the Extension onboarding

![1](https://github.com/woocommerce/google-listings-and-ads/assets/17420811/71d4aacf-fed6-4f5f-8275-6092c5ad61fc)

#### 📷 Event tracking when clicking on button to open the claim page of Google Ads account in the Extension onboarding

![2](https://github.com/woocommerce/google-listings-and-ads/assets/17420811/09a486ad-9595-42d2-afb1-23e891ff360b)

#### 📷 Event tracking when clicking on button to create Google Ads account in the Ads onboarding

![3](https://github.com/woocommerce/google-listings-and-ads/assets/17420811/0cd0403a-9d3a-4852-b841-4d19b7776da4)

#### 📷 Event tracking when clicking on button to open the claim page of Google Ads account in the Ads onboarding

![4](https://github.com/woocommerce/google-listings-and-ads/assets/17420811/37225048-6578-4fde-8b8e-097352399177)


### Detailed test instructions:

1. Open the Console tab in the browser DevTool, and run `localStorage.setItem( 'debug', 'wc-admin:*' )` to enable tracking debugging mode.
2. Disconnect Google Ads account if there is a connected one.
3. Go to step 1 of the **Extension onboarding**.
4. Create a new Google Ads account.
5. Check if the `gla_ads_account_create_button_click` event is sent with `context` and `step`
6. Click the "Claim account" button to open the invitation link in a pop-up window for the newly created Google Ads account. 
   - It doesn't need to proceed with the claim process here
7. Check if the `gla_ads_account_create_button_click` event is sent with `context` and `step`.
8. Disconnect Google Ads account if there is a connected one.
9. Go to step 1 of the **Ads onboarding**.
10. Repeat test steps 4-7

### Changelog entry

> Tweak - Adjust event tracking for the creating and claim buttons of Google Ads account.
